### DISCUSSION TITLE
Antd without less

### DISCUSSION THREAD

**Discussion by StKostyantin** [ant-design/ant-design#30413] [Q&A]
Are there any plans to get rid of less?

**Comment by jueinin** [comment]
why get rid of less?

**Reply by StKostyantin** [reply]
Because less is not actively used in projects and this greatly complicates integration. For example, in a PR created to solve the problem of integrating antd with nextjs. The developer answers why this PR will not be confirmed.
https://github.com/vercel/next.js/pull/23185#pullrequestreview-647986468

Antd is the best solution for today, but less slows down its development.

**Reply by jueinin** [reply]
Yeah you are right. Less is less popular than sass, and create react app cli does not support less natively too.

**Reply by StKostyantin** [reply]
That's it! Because of this, it is necessary to customize the webpack and this slows down development, slows down the webpack, and it also sometimes increases the build at the output

**Reply by Gervwyk** [reply]
Totally agree with this. Also makes customization very hard since all customization must be loaded at build time.

**Comment by Gervwyk** [comment]
What advantages does the less dependency add which cannot be solved with css variables?

**Reply by StKostyantin** [reply]
To reduce the size of the bundle and not write your own styles on top of antd, it is possible to substitute variables.
https://ant.design/docs/react/customize-theme

But this is difficult to do when less is not supported. Because of this, you have to configure the webpack and because of this, development productivity is lost.

**Reply by StKostyantin** [reply]
You can import regular styles and write your own styles on top of them, but this will increase the bundle and create cascading problems.

**Reply by Gervwyk** [reply]
I agree, less introduces problems like build complexities. But if antd were to not use less, and purely use css variables for theme customization, this would enable first class support for dynamic theme changes, instead of hacky workaround like [these](https://medium.com/@mzohaib.qc/ant-design-dynamic-runtime-theme-1f9a1a030ba0).

**Reply by StKostyantin** [reply]
This setup is much easier with sass or even css-in-js. Due to the large options for customization solutions.

**Reply by momesana** [reply]
Less is a royal pain and antd would benefit from getting rid of it. That's a colossal task however and as an interim solution it would be enough to just expose all relevant less variables as CSS variables as is already being done with the [primary colors ](https://ant.design/docs/react/customize-theme-variable) and similar variables. Notably background colors are missing though which makes it impossible to support dark themes.

**Reply by amille44420** [reply]
For customization I would indeed rather be looking into playing with a context (like React Context) which does not imply thinking about how to customize the build itself, but rather fully rely on the runtime. It would remove plenty loaders from webpack as well and enhance the customization making it fully dynamic. 

Basically on my projects so far, `antd` is the only one wish requires me dealing with LESS/CSS. 

I do prefer working with `antd` for the library it provides but for this single reason I have many developers wishing to use `material-ui` instead as it's easier to deal with customizations thanks to JSS principles.
